alf/algorithms/lagrangian_algorithm.py

```diff
@@ -0,0 +1,134 @@
+# Copyright (c) 2021 Horizon Robotics and ALF Contributors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Lagrangian Algorithm."""
+
+import numpy as np
+import gin
+
+import torch
+import torch.nn as nn
+from torch.nn import functional as F
+
+import alf
+from alf.algorithms.algorithm import Algorithm
+from alf.data_structures import namedtuple, AlgStep
+from alf.nest import nest
+from alf.tensor_specs import TensorSpec
+from alf.utils import tensor_utils
+
+LagInfo = namedtuple("LagInfo", ["rollout_reward"], default_value=())
+
+
+@alf.configurable(blacklist=["reward_spec"])
+class LagrangianAlgorithm(Algorithm):
```

Comment:
"Lagrangian" seems to general here.  This works only on reward weight.

Reply:
changed